### PR TITLE
Document no_expect test harness helpers and fix context detection

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Document the `no_expect_outside_tests` UI harness helpers and dependency-rlib test support, and extend the developer guide's regression-infrastructure section with the harness workflow details.

Also tighten `collect_context` so function ancestors tagged via configured `additional_test_attributes` are treated as test context during summarisation.

## Why

The crate had documentation coverage warnings around the UI harness helpers and rlib-selection tests, and the maintainer guide did not explain when these harness helpers should be used.

The context-collection path also needed to ensure configured test-like attributes are recognised while walking HIR ancestors so async test harness regressions are classified correctly.

## Impact

- Raises Rustdoc coverage for the affected helper types and functions.
- Documents the harness parameter structs, driver functions, and rlib resolution path for maintainers.
- Keeps additional test attributes active during context summarisation for `no_expect_outside_tests`.

## Validation

- `cargo test -p no_expect_outside_tests --features dylint-driver`
- `make markdownlint`
- `make nixie`
- `make fmt`
- `make check-fmt`
- `make lint`
- `make test`

## Summary by Sourcery

Document the no_expect UI test harness helpers and refine test context detection for additional test attributes.

Bug Fixes:
- Ensure HIR ancestor traversal treats functions tagged via configured additional_test_attributes as test context when collecting context.

Documentation:
- Expand rustdoc for harness parameter structs, driver functions, and rlib resolution helpers in lib_ui_tests.rs and dependency_rlib_tests.rs.
- Extend the developer guide with a detailed description of the UI test harness workflow, parameter structs, rlib resolution behaviour, and camino usage.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### crates/no_expect_outside_tests/src/driver/tests.rs (1)

`157-174`: _⚠️ Potential issue_ | _🟡 Minor_
**Update the coverage note to match the new heuristic.**
Rewrite this block. It still describes `is_test_named_module`-style coverage and `mod test/tests` detection, but the implementation now relies on arbitrary `cfg(test)` ancestry plus the Tokio/path-loaded regressions added in this PR.

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai pause

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Individual Comments

### Comment 1
<location path="crates/no_expect_outside_tests/src/dependency_rlib_tests.rs" line_range="60" />
<code_context>
+/// Creates `artifacts` inside `directory`, sets their modification times, then
+/// invokes `dependency_rlib` and returns both the expected and selected paths
+/// for the caller to compare.
 fn resolve_dependency_rlib_selection(
     directory: TemporaryDirectory,
     artifacts: &[ArtifactSpec<'_>],
</code_context>
<issue_to_address>
**suggestion (testing):** Add a negative test case for when no matching rlib is present

Current coverage only exercises the successful selection path. Please add a test that uses an empty directory (or one with only non-matching files) and asserts that `dependency_rlib` returns an `Err` with the expected message shape, so the error path is exercised and the “no suitable artefact found” case is verified.

Suggested implementation:

```rust
/// rstest fixture that creates a uniquely named temporary directory for a
/// selection test.
#[fixture]
fn selection_directory() -> TemporaryDirectory {
    TemporaryDirectory::new("selection")
}

/// Ensures that `dependency_rlib` returns an error when no matching `rlib` is present.
///
/// This exercises the error path so the “no suitable artefact found” case is verified.
#[rstest]
fn dependency_rlib_returns_err_when_no_matching_rlib_present(
    selection_directory: TemporaryDirectory,
) {
    // The `selection_directory` is empty here (no artifacts created).

    let result = dependency_rlib(
        selection_directory.path(),
        // Adjust these parameters to match the real `dependency_rlib` signature:
        // e.g. crate name, target, profile, etc.
    );

    assert!(
        result.is_err(),
        "expected dependency_rlib to return an error when no matching rlib is present",
    );

    let err_msg = result.unwrap_err().to_string();
    assert!(
        err_msg.contains("no suitable artefact found")
            || err_msg.contains("no suitable artifact found"),
        "unexpected error message, got: {err_msg}",
    );
}

```

1. Update the call to `dependency_rlib` in the new test to pass the correct arguments for your actual `dependency_rlib` function signature (crate name, target triple, profile, etc. as required).
2. If `TemporaryDirectory` does not expose a `.path()` method returning `&Path`, replace `selection_directory.path()` with the appropriate accessor.
3. If the concrete error type produced by `dependency_rlib` is not `impl Display` or does not stringify cleanly, adjust `let err_msg = ...` accordingly (e.g. use `&err.msg` or similar).
4. Align the asserted error text with the real message used in `dependency_rlib` (if it always uses “artefact” vs “artifact”, you can simplify the assertion to check only the exact string).
</issue_to_address>

### Comment 2
<location path="docs/developers-guide.md" line_range="826" />
<code_context>
+| `run_fixture_under_test_harness` | Prepares a fixture directory, assembles rustc flags (including extern-crate links), and delegates to `run_test_runner`.                |
+| `run_fixture_harness_test`       | End-to-end driver: calls `run_with_runner` and panics with a labelled message on failure. Used by parametrised `#[rstest]` test cases. |
+
+#### Dependency rlib resolution
+
+When a regression requires a real external crate (e.g. `tokio`), the harness
</code_context>
<issue_to_address>
**suggestion (review_instructions):** The acronym “rlib” is introduced without being defined on first use, which violates the acronym-definition guideline.

Consider expanding this heading or the first sentence beneath it to define “rlib”, for example: “Dependency Rust library (rlib) resolution”. This will satisfy the requirement to define uncommon acronyms on first use.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: docs/developers-guide.md [hunk 1]
```diff
@@ -795,6 +795,55 @@ Keep the regression split aligned with that compiler boundary:
...
+  as `--extern` dependencies).
+
+#### Harness driver functions
+
+| Function                         | Purpose                                                                                                                                |
+| -------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------- |
+| `run_example_under_test_harness` | Runs one example target under the dylint UI test harness using the flags in the `ExampleHarnessRun` spec.                              |
+| `run_fixture_under_test_harness` | Prepares a fixture directory, assembles rustc flags (including extern-crate links), and delegates to `run_test_runner`.                |
+| `run_fixture_harness_test`       | End-to-end driver: calls `run_with_runner` and panics with a labelled message on failure. Used by parametrised `#[rstest]` test cases. |
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (4 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing | ⚠️ Warning | The new `has_test_ancestry` helper function lacks direct unit test coverage; whilst behavioural coverage exists through UI tests, isolated unit tests for the three conditions are missing. | Add unit tests to `crates/no_expect_outside_tests/src/context/tests.rs` for `has_test_ancestry` verifying each condition independently and in combination using `#[rstest]` parameterised tests. |
| User-Facing Documentation | ⚠️ Warning | The pull request changes when `additional_test_attributes` take effect during ancestor context detection, expanding their scope beyond direct annotations. This user-facing behavioural change is not documented in `docs/users-guide.md`. | Update `docs/users-guide.md` to document that `additional_test_attributes` now apply to ancestor functions during context detection, with a concrete example showing how parent function annotations mark nested code as test context. |
| Developer Documentation | ⚠️ Warning | The PR documents UI test harness helpers but fails to document the three-way test-context detection mechanism in `has_test_ancestry()` and how `additional_test_attributes` enable custom test frameworks to be recognised during ancestor scanning. | Document the test-context detection architecture in the developers-guide: describe the three-way detection mechanism, explain how `additional_test_attributes` work with ancestor scanning, clarify `collect_context` and `summarise_context` roles, and document why this pattern enables lints to respect user-defined test markers across the entire ancestry chain. |
| Module-Level Documentation | ⚠️ Warning | Three pub(crate) functions in context.rs lack required documentation: collect\_context, summarise\_context, and is\_cfg\_test\_attribute have no preceding doc comments. | Add comprehensive doc comments to the three undocumented pub(crate) functions, including purpose, parameters, return types, and examples. |
